frontend/rust-lib/flowy-document/src/event_handler.rs

```diff
@@ -401,3 +406,125 @@ pub(crate) async fn convert_data_to_json_handler(
 
   data_result_ok(ConvertDataToJsonResponsePB { json: result })
 }
+
+// Handler for uploading a file
+// `workspace_id` and `file_name` determines file identity
+pub(crate) async fn upload_file_handler(
+  params: AFPluginData<UploadFileParamsPB>,
+  manager: AFPluginState<Weak<DocumentManager>>,
+) -> DataResult<UploadedFilePB, FlowyError> {
+  let manager = upgrade_document(manager)?;
+  let file_service = manager
+    .get_file_storage_service()
+    .upgrade()
+    .ok_or(FlowyError::internal().with_context("The file storage service is already dropped"))?;
+
+  let mut file = tokio::fs::File::open(&params.local_file_path).await?;
+  let mut content = Vec::new();
+  let n = file.read_to_end(&mut content).await?;
+  info!("read {} bytes from file: {}", n, &params.local_file_path);
+
+  let hash = fxhash::hash(&content);
+  let mime = mime_guess::from_path(&params.local_file_path).first_or_octet_stream();
+
+  let url = file_service
+    .get_object_url(ObjectIdentity {
+      workspace_id: params.workspace_id.to_owned(),
+      file_id: hash.to_string(),
+    })
+    .await?;
+
+  // let the upload happen in the background
+  let url_clone = url.clone();
```

Comment:
I think the `cloned_url` would be the more conventional choice in Rust  In most cases, But url_clone is also ok

Reply:
changed